### PR TITLE
context: Do not handle hotplug events after dispose

### PR DESCRIPTION
**Pull request by benzea**
Hotplug events may be queued at any time. They kept the object alive,
but an explicit expose would mean that the event would be processed
after dispose but before finalize.

Change the code to instead use a weak pointer. This causes the handling
to be aborted if the object has been disposed in the meantime.

We could also simply check whether the internal state is acceptable at
the time (e.g. check thread_event_run). However, using weak pointers has
the advantage that we'll dispose/finalize earlier if the API user is just
doing an unref, making the API more predictable.

This fixes commit 24934619a2ad ("Fix hotplug after threaded device
removal") which in turn was a fix for commit 4b52b0cd27fd ("context: Fix
hotplug handling and signal emission during enumerate")

### CODE REVIEW

**Comment by benzea** [comment]
So, the test verifies that the fix is correct.

But, it crashes in `g_usb_context_find_by_bus_address` as `g_usb_context_add_device` calls it. The funny thing is, the libfprint test case does not crash, which points to `priv->done_enumerate` being `FALSE`. I don't have a good explanation for that, but I suppose that might be a memory corruption, even if it is quite consistent.

So, not 100% certain this fixes the random libfprint failure. But, pretty certain the bugfix is correct.

**Comment by hughsie** [comment]
I'm a bit confused; `libusb_hotplug_deregister_callback()` gets called, we join the thread, and then the callback gets triggered?

File: gusb/gusb-umockdev-test.c [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+	devices = g_usb_context_get_devices (fixture->ctx);
+	g_assert_cmpint (devices->len, ==, 0);
+	g_assert_cmpint (events, ==, 0);
+	g_clear_pointer (&devices, g_ptr_array_unref);
+
+	/* The idle handler is pending, we dispose our context reference */
+	g_object_run_dispose (G_OBJECT (fixture->ctx));
+
+	/* Run mainloop, which causes the event to be processed. */
+	while (g_main_context_iteration (NULL, FALSE)) { }
```
Comment (benzea) [inline gusb/gusb-umockdev-test.c:182]: This is where the crash happens without the fix.

 * `test_fixture_add_canon` generates the hotplug event
 * `g_usb_context_hotplug_cb` is called in libusb event thread (the sleep ensures this happens)
 * We dispose the GUsbContext
 * mainloop runs: `g_usb_context_idle_hotplug_cb` runs and calls `g_usb_context_add_device`

Comment (benzea) [reply gusb/gusb-umockdev-test.c:182]: Yeah, it doesn't prevent dispatching, it should prevent issues.

Your solution looks better to me, just one typo. I'll file a new MR for the tests (or you can just cherry-pick them).

**Comment by benzea** [comment]
Obsoleted by #58.

**Pull request closed without merging**